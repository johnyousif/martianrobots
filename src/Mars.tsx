import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import styled from 'styled-components';
import { createGetOutputMovementsRequestedAction } from './redux/actionCreators';
import { RootState } from './redux/root';
import {
  getOutputMovements,
  getOutputMovementsError,
} from './redux/selectors/movement';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const InstructionInput = styled.textarea`
  width: 50vw;
  height: 25vh;
  resize: none;
  margin-bottom: 20px;
  padding: 10px;
`;

const Output = styled.div`
  width: 50vw;
  height: 25vh;
  margin-top: 20px;
  padding: 10px;
  background: #000000;
  color: #ffffff;
  white-space: pre-wrap;
`;

const Error = styled.p`
  font-size: 20px;
  font-weight: bold;
  margin-top: 20px;
  color: red;
`;

interface IStateProps {
  outputMovements: string;
  outputMovementsError?: string;
}

interface IDispatchProps {
  getOutputMovements: (inputMovements: string) => void;
}

const Mars: React.FC<IStateProps & IDispatchProps> = ({
  outputMovements,
  outputMovementsError,
  getOutputMovements,
}) => {
  const [robotInstructions, setRobotInstructions] = useState<string>('');

  const onChangeRobotInstructions = (
    e: React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    setRobotInstructions(e.target.value);
  };

  const onClickExecuteRobotInstructions = () => {
    getOutputMovements(encodeURI(robotInstructions));
  };
  return (
    <Container>
      <InstructionInput
        placeholder="Enter input instructions here..."
        onChange={onChangeRobotInstructions}
        value={robotInstructions}
      />
      <button
        disabled={!robotInstructions.length}
        onClick={onClickExecuteRobotInstructions}
      >
        Execute robot instructions
      </button>
      <Output>{outputMovements || 'Output will appear here...'}</Output>
      {outputMovementsError && <Error>{outputMovementsError}</Error>}
    </Container>
  );
};

const mapStateToProps = (state: RootState): IStateProps => ({
  outputMovements: getOutputMovements(state),
  outputMovementsError: getOutputMovementsError(state),
});

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => ({
  getOutputMovements: (inputMovements: string) =>
    dispatch(createGetOutputMovementsRequestedAction(inputMovements)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Mars);
