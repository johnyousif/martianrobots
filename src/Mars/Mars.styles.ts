import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const InstructionInput = styled.textarea`
  width: 50vw;
  height: 25vh;
  resize: none;
  margin-bottom: 20px;
  padding: 10px;
`;

export const Output = styled.div`
  width: 50vw;
  height: 25vh;
  margin-top: 20px;
  padding: 10px;
  background: #000000;
  color: #ffffff;
  white-space: pre-wrap;
`;

export const OutputError = styled.p`
  font-size: 20px;
  font-weight: bold;
  margin-top: 20px;
  color: red;
`;
