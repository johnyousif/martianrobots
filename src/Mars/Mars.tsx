import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import {
  Container,
  InstructionInput,
  Output,
  OutputError,
} from './Mars.styles';
import { createGetOutputMovementsRequestedAction } from '../redux/actionCreators';
import { RootState } from '../redux/root';
import {
  getOutputMovements,
  getOutputMovementsError,
} from '../redux/selectors/movement';

interface IStateProps {
  outputMovements: string;
  outputMovementsError?: string;
}

interface IDispatchProps {
  getOutputMovements: (inputMovements: string) => void;
}

export const Mars: React.FC<IStateProps & IDispatchProps> = ({
  outputMovements,
  outputMovementsError,
  getOutputMovements,
}) => {
  const [robotInstructions, setRobotInstructions] = useState<string>('');

  const onChangeRobotInstructions = (
    e: React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    setRobotInstructions(e.target.value);
  };

  const onClickExecuteRobotInstructions = () => {
    getOutputMovements(encodeURI(robotInstructions));
  };
  
  return (
    <Container>
      <InstructionInput
        className="instruction-input"
        placeholder="Enter input instructions here..."
        onChange={onChangeRobotInstructions}
        value={robotInstructions}
      />
      <button
        disabled={!robotInstructions.length}
        onClick={onClickExecuteRobotInstructions}
      >
        Execute robot instructions
      </button>
      <Output>{outputMovements}</Output>
      {outputMovementsError && (
        <OutputError className="output-error">
          {outputMovementsError}
        </OutputError>
      )}
    </Container>
  );
};

const mapStateToProps = (state: RootState): IStateProps => ({
  outputMovements: getOutputMovements(state),
  outputMovementsError: getOutputMovementsError(state),
});

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => ({
  getOutputMovements: (inputMovements: string) =>
    dispatch(createGetOutputMovementsRequestedAction(inputMovements)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Mars);
