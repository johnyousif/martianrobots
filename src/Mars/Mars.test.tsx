import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Mars } from './Mars';

describe('Mars tests', () => {
  it('should render without errors', () => {
    expect(
      toJson(
        shallow(<Mars outputMovements="" getOutputMovements={jest.fn()} />)
      )
    ).toMatchSnapshot();
  });

  it('should display user input', () => {
    const wrapper = shallow(
      <Mars outputMovements="" getOutputMovements={jest.fn()} />
    );
    let instructionInputComponent = wrapper.find('.instruction-input');

    expect(instructionInputComponent.text()).toEqual('');

    instructionInputComponent.simulate('change', { target: { value: 'xyz' } });
    instructionInputComponent = wrapper.find('.instruction-input');

    expect(instructionInputComponent.html().includes('xyz')).toBeTruthy();
  });

  it('should call output api', () => {
    const getOutputMovements = jest.fn();
    const wrapper = shallow(
      <Mars outputMovements="" getOutputMovements={getOutputMovements} />
    );
    wrapper.find('button').simulate('click');

    expect(getOutputMovements).toHaveBeenCalledTimes(1);
  });

  it('should display output error', () => {
    const wrapperWithoutError = shallow(
      <Mars outputMovements="" getOutputMovements={jest.fn()} />
    );

    const wrapperWithError = shallow(
      <Mars
        outputMovements=""
        outputMovementsError="error"
        getOutputMovements={jest.fn()}
      />
    );

    let outputErrorComponent = wrapperWithoutError.find('.output-error');

    expect(outputErrorComponent.exists()).toBeFalsy();

    outputErrorComponent = wrapperWithError.find('.output-error');

    expect(outputErrorComponent.exists()).toBeTruthy();
    expect(outputErrorComponent.text()).toEqual('error');
  });
});
