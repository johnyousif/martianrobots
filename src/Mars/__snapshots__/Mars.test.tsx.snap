// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mars tests should render without errors 1`] = `
<styled.div>
  <styled.textarea
    className="instruction-input"
    onChange={[Function]}
    placeholder="Enter input instructions here..."
    value=""
  />
  <button
    disabled={true}
    onClick={[Function]}
  >
    Execute robot instructions
  </button>
  <styled.div />
</styled.div>
`;
